bluemira/equilibria/fem_fixed_boundary/equilibrium.py

```diff
@@ -0,0 +1,256 @@
+# License along with bluemira; if not, see <https://www.gnu.org/licenses/>.
+
+"""Fixed boundary equilibrium class"""
+import numpy as np
+
+from bluemira.base.look_and_feel import bluemira_debug, bluemira_print, bluemira_warn
+from bluemira.equilibria.fem_fixed_boundary.fem_magnetostatic_2D import (
+    FemGradShafranovFixedBoundary,
+)
+from bluemira.equilibria.fem_fixed_boundary.transport_solver import (
+    PlasmodTransportSolver,
+)
+from bluemira.equilibria.fem_fixed_boundary.utilities import (
+    calculate_plasma_shape_params,
+    plot_profile,
+)
+from bluemira.mesh import meshing
+from bluemira.mesh.tools import import_mesh, msh_to_xdmf
+
+
+def solve_plasmod_fixed_boundary(
+    builder_plasma,
+    plasmod_options,
+    gs_options,
+    delta95_t,
+    kappa95_t,
+    lcar_mesh=0.15,
+    max_iter=30,
+    iter_err_max=1e-5,
+    relaxation=0.2,
+    plot=False,
+    verbose=False,
+):
+    """
+    Solve the plasma fixed boundary problem using delta95 and kappa95 as target
+    values and iterating on PLASMOD to have consistency with pprime and ffprime.
+
+    Parameters
+    ----------
+    builder_plasma: Builder
+        Plasma poloidal cross section builder object
+    plasmod_options: dict
+        Set of options used to set up and run PLASMOD
+    gs_options: dict
+        Set of options used to set up and run the FemGradShafranovFixedBoundary
+    delta95_t: float
+        Target value for delta at 95%
+    kappa95_t: float
+        Target value for kappa at 95%
+    lcar_mesh: float
+        Value of the characteristic length used to generate the mesh to solve the
+        Grad-Shafranov problem
+    max_iter: int
+        Maximum number of iteration between Grad-Shafranov and PLASMOD
+    iter_err_max: float
+        Convergence maximum error to stop the iteration
+    relaxation: float
+        Iteration relaxing factor
+    plot: bool
+        Whether or not to plot
+    verbose: bool
+        Whether or not to print
+
+    Notes
+    -----
+    This function directly modifies the parameters of builder_plasma
+    """
+    n_iter = 0
+    delta_95 = delta95_t
+    kappa_95 = kappa95_t
+
+    while n_iter < max_iter:
+        # source string to be used in changed parameters
+        source = f"from equilibrium iteration {n_iter}"
+
+        # build the plasma x-z cross-section and get its volume
+        plasma = builder_plasma.build_xz().get_component("xz").get_component("LCFS")
+        lcfs = plasma.shape
+        plasma_volume = 2 * np.pi * lcfs.center_of_mass[0] * lcfs.area
+
+        if plot:
+            plasma.plot_options.show_faces = False
+            plasma.plot_2d(show=True)
+
+        kappa_u = builder_plasma.params.get_param("kappa_u")
+        kappa_l = builder_plasma.params.get_param("kappa_l")
+        delta_u = builder_plasma.params.get_param("delta_u")
+        delta_l = builder_plasma.params.get_param("delta_l")
+
+        kappa = (kappa_u + kappa_l) / 2
```

Comment:
Not sure where to go about this (probably in the input rather than in here), but one way of speeding up convergence would be to set the starting kappa_u and kappa_l such that they equal 1.12 * kappa_95, which is a rule of thumb used in PROCESS based on some FIESTA scans. Of course, there is the awkward decision of how to split between kappa_u and kappa_l, but that is not going to go away for SNs using the JohnerLCFS parameterisation.